Fix & improve typing messages

AltGr opened the pull request:
This attempts to improve a few error messages, and fixes a bug where typing was
done unexpectedly lazily.

AltGr commented:
NOTE: is on top of #340 and should be rebased once that is merged.

denismerigoux commented on compiler/desugared/ast.ml:
```diff
@@ -40,7 +40,7 @@ module LabelSet : Set.S with type elt = LabelName.t = Set.Make (LabelName)
   type t =
     | Var of ScopeVar.t * StateName.t option
-    | SubScopeVar of SubScopeName.t * ScopeVar.t
+    | SubScopeVar of SubScopeName.t * ScopeVar.t * Pos.t
```
To what does the `Pos.t` points to ? And why only for `SubScopeVar` and not for `Var` ?

AltGr replied:
`ScopeVar.t` / `StateName.t` already contain a position; `SubScopeName.t` as well. But in this case, the position of the dot `subscope.var` can be different from both the above.

denismerigoux commented on compiler/desugared/ast.ml:
```diff
@@ -49,35 +49,36 @@ module ScopeDef = struct
     | Var (x, None) -> Marked.get_mark (ScopeVar.get_info x)
     | Var (_, Some sx) -> Marked.get_mark (StateName.get_info sx)
-    | SubScopeVar (x, _) -> Marked.get_mark (SubScopeName.get_info x)
+    | SubScopeVar (_, _, pos) -> pos
```
Ok I see it's for error messages.

denismerigoux commented on tests/test_typing/bad/err2.catala_en:
```diff
@@ -23,7 +23,7 @@ Error coming from typechecking the following expression:
    + 
 
-Type Structure array coming from expression:
+Type any[10] array coming from expression:
```
So this is not an improvement. The union find representative chosen should not be a type variable if there is a non-type-variable representative.

denismerigoux commented on tests/test_typing/bad/err3.catala_en:
````diff
@@ -12,5 +12,61 @@ scope S:
+#return code 255#
+```
+
+A bug with delayed typechecking can make the above accept the error while it is
````
What is that ?

AltGr replied:
Just a comment on why the test is needed. Should it rather be in a ``catala block ?

AltGr replied:
Here we get the same error twice, which is correct behaviour. If you look at the sate before the fix the test above answers "typing successful": https://github.com/CatalaLang/catala/pull/341/commits/498429e4b7765634c05e538dc96e28524e6e1410#diff-bde130b0c71ac3cb3a70fd86df22d09dc60f2b704391cf2c7bf2fb09c5613dfbL15

denismerigoux reviewed (commented):
Some things have improved indeed but type variables appearing in error messages is a no-go for me

AltGr commented:
@denismerigoux a41de29293b should make the message better, ready for review.

denismerigoux commented on build_system/clerk_driver.ml:
```diff
@@ -885,7 +885,7 @@ let driver
                    ctx.all_test_builds));
           let ninja_cmd =
-            "ninja -f " ^ ninja_output ^ " " ^ ninja_flags ^ " test"
+            "ninja -k 0 -f " ^ ninja_output ^ " " ^ ninja_flags ^ " test"
```
Doesn't this reintroduced what was fixed by #337  ?

denismerigoux reviewed (approved):
Thanks! Sounds good except for this `err3` delayed thing which looks bad to me.

AltGr commented:
> Thanks! Sounds good except for this err3 delayed thing which looks bad to me.

Actually this fixes it, I introduced that regression when I added the type annotations in the AST I guess.
